stock_archive_constraint/models/product_product.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2020 Tecnativa - Víctor Martínez
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, models
+from odoo.exceptions import ValidationError
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    @api.constrains("active")
+    def _check_active_stock_archive_constraint_stock_quant(self):
+        res = self.env['stock.quant'].search(
+            [
+                ('product_id', 'in', self.ids),
+                ('quantity', '!=', 0.0)
+            ], limit=1
+        )
+        if res:
+            raise ValidationError(
+                _("It is not possible to archive product '%s' which has "
+                  "associated quants." % res.product_id.display_name)
+            )
```

Comment:
Example:
1. Manufacture product
2. Sold all product qty
3. Archive product
What has to happen?
Allow archive or not?

Reply:
Yes, the quant search should be limited to internal locations, not virtual ones.

Reply:
I think transit location is important too

Reply:
Changed it now.